File: algorithms/backtracking/subset-sum/c/subset_sum.c
static int backtrack(int arr[], int n, int index, int remaining) {
    if (remaining == 0) {
        return 1;
    }
    if (index >= n) {
        return 0;
    }
    /* Include arr[index] */
    if (backtrack(arr, n, index + 1, remaining - arr[index])) {
        return 1;
    }
    /* Exclude arr[index] */
    if (backtrack(arr, n, index + 1, remaining)) {
        return 1;
    }
    return 0;
}

int subset_sum(int arr[], int n, int target) {
    return backtrack(arr, n, 0, target);
}

File: algorithms/backtracking/subset-sum/kotlin/SubsetSum.kt
fun subsetSum(arr: IntArray, target: Int): Int {
    return if (backtrack(arr, 0, target)) 1 else 0
}

private fun backtrack(arr: IntArray, index: Int, remaining: Int): Boolean {
    if (remaining == 0) return true
    if (index >= arr.size) return false

    // Include arr[index]
    if (backtrack(arr, index + 1, remaining - arr[index])) return true

    // Exclude arr[index]
    if (backtrack(arr, index + 1, remaining)) return true

    return false
}
